### DISCUSSION TITLE
Mount model to custom form

### DISCUSSION THREAD

**Discussion by studiowizjo** [filamentphp/filament#3494] [Help] [answered]
I would like to have custom form which I can reuse to both create and edit comment, but I got problems with mounting model to it. Data is not visible in form while editing and when dumping after save.

Here is my code:

```php

public $manageComment;

protected function getForms(): array
{
    return [
        'commentForm' => $this->getCommentForm(),
    ];
}

public function openCommentModal(int $taskId, ?int $commentId = null)
{
    $this->manageComment = $commentId
        ? Comment::find($commentId)
        : new Comment(['task_id' => $taskId]);

    $this->dispatchBrowserEvent('open-modal', ['id' => 'modalComment']);
}

public function getCommentForm()
{
    return $this->makeForm()
        ->schema([
            Grid::make()
                ->schema([
                    //fields
                ])->columns(3)
        ])
        ->model($this->manageComment);
}

public function saveCommentForm()
{
    dd($this->getCommentForm()->model);
}
```

and I open it in view like this:

```blade
<a wire:click.prevent="openCommentModal(7921)" href="">Add comment to task #7921</a>
<a wire:click.prevent="openCommentModal(7921, 14007)" href="">Edit comment #14007</a>
```

**Comment by danharrin** [comment, marked as answer]
You can't mount a non-existant model object to a Filament form, in that case you would need to pass in a Model::class instead:

```php
public $manageComment;
public $manageTask;

protected function getForms(): array
{
    return [
        'commentForm' => $this->getCommentForm(),
    ];
}

public function openCommentModal(int $taskId, ?int $commentId = null)
{
    $this->manageTask = Task::find($taskId);
    $this->manageComment = Comment::find($commentId);

    $this->dispatchBrowserEvent('open-modal', ['id' => 'modalComment']);
}

public function getCommentForm()
{
    return $this->makeForm()
        ->schema([
            Grid::make()
                ->schema([
                    //fields
                ])->columns(3)
        ])
        ->model($this->manageComment ?? Comment::class);
}

public function saveCommentForm()
{
    $data = $this->commentForm->getState();

    if ($this->manageComment) {
        $this->manageComment->update($data);

        return;
    }

    $comment = new Comment();
    $comment->fill($data);
    $comment->task()->associate($this->manageTask);
    $comment->save();
}
```

**Reply by studiowizjo** [reply]
Unfortunately, this doesn`t solve my issue. When I click:

```blade
<a wire:click.prevent="openCommentModal(7921, 14007)" href="">Edit comment #14007</a>
```

I expect form should be filled with comment data, but it is not.

Here are some additional details:

My (simplified) form fields:

```php
TextInput::make('task_id')
    ->prefix('#')
    ->required(),
TinyEditor::make('content')
    ->columnSpan(3),
```

My resource is custom and extends ```ManageRecords```:

```php
class ProjectTasks extends ManageRecords
```

I use custom view which (again, simplified) looks like this:

```blade
<x-filament::page class="filament-resources-list-records-page">

    {{-- <a @click.prevent="$dispatch('open-modal', { id: 'modalComment'})" href="">open alpine</a> --}}
    <a wire:click.prevent="openCommentModal(7921)" href="">Add comment to task #7921</a>
    <a wire:click.prevent="openCommentModal(7921, 14007)" href="">Edit comment #14007</a>

    <x-filament::modal id="modalComment" width="3xl">
        <x-slot name="header">
            Add comment
        </x-slot>

        <x-slot name="heading"></x-slot>
        <x-slot name="subheading"></x-slot>

        {{ $this->commentForm }}

        <x-slot name="footer">
            <x-filament::modal.actions class="filament-modal-actions flex flex-wrap items-center gap-4 rtl:space-x-reverse">
                <x-filament::button x-on:click="isOpen = false" color="secondary">
                    {{ __('filament::resources/pages/edit-record.modals.delete.buttons.cancel.label') }}
                </x-filament::button>

                <x-filament::button wire:click="saveCommentForm" color="primary">
                    Save
                </x-filament::button>
            </x-filament::modal.actions>
        </x-slot>

    </x-filament::modal>

</x-filament::page>
```

**Reply by danharrin** [reply]
```php
public function openCommentModal(int $taskId, ?int $commentId = null)
{
    $this->manageTask = Task::find($taskId);
    $this->manageComment = $this->manageTask->comments()->find($commentId);

    $this->commentForm->fill($this->manageComment?->attributesToArray());

    $this->dispatchBrowserEvent('open-modal', ['id' => 'modalComment']);
}
```

**Reply by studiowizjo** [reply]
I thought that:

```php
->model($this->manageComment ?? Comment::class)
```

is responsible also for filling the form.


Anyway it is getting closer, but there is still a problem. Console shows error:

```
Livewire Entangle Error: Livewire property 'content' cannot be found
```

and `content` form field doesn`t loads.

Normally (on regular Livewire component), I think, this should be fixed by adding key to  ```$rules``` array or typehinting ```public ?Comment $manageComment;```, but it is not the case.

**Reply by danharrin** [reply]
Ah right. You want a `public $commentData` property, and pass `->statePath('commentData')` in the `getCommentForm()` method.

By default, Filament forms need a public property for every form field (hence why the error is there as Livewire is searching for a `$content` property), but by passing a state path, Filament saves all data within the `$commentData` array property.

**Reply by studiowizjo** [reply]
Perfect, thank you very much =).

**Reply by studiowizjo** [reply]
One more thing I got problems with. If one of my form field is:

```php
SpatieMediaLibraryFileUpload::make('attachments')
    ->collection('tasks'),
```

what would be the easiest way to save uploaded files? I`m tearing my hairs to figure this out =/. 

Do I need to write my own logic or there is any method I could benefit from?

**Reply by danharrin** [reply]
```php
public function saveCommentForm()
{
    // ...

    $this->commentForm->model($comment)->saveRelationships();
}
```

If a model instance is passed to the form when `getState()` is called, the relationships are saved automatically, so the edit form is already handled. However, since the comment model could also be created, you need to add this extra line to save the relationships after you save the comment. It's in the docs.

**Reply by studiowizjo** [reply]
I read that at docs and found it also in source, but this was not working for me. I tried to implement it very similiar to [example from docs](https://filamentphp.com/docs/2.x/forms/getting-started#saving-field-relationships-manually) and your tips above, but Im missing something.

Here is a current example I am working with. This time I am editing existing `Task`, so model is always avalaible.

The problem is that all properties are binded to form, apart from `attachments`. Also when saving, uploaded files are not attached.

I use the same form schema in `TaskResource` `form()` method and from there attachments are saved correctly, so the relation and collection on field is the same. 

```php
public Task $task;

public $taskData = [
    'id' => null,
    'project_id' => null,
    'url' => null,
    'description' => null,
    'attachments' => null,
    //...
];

public function openTaskModal(Task $task)
{
    $task->load('media');
    $this->task = $task;

    $this->taskForm->model($this->task);
    $this->taskForm->fill($this->task->toArray());

    $this->dispatchBrowserEvent('open-modal', ['id' => 'modalTask']);
}

public function getTaskForm()
{
    return $this->makeForm()
        ->schema([
            Grid::make()
                ->schema([
                    TextInput::make('url')
                        ->required(),

                    Select::make('project_id')
                        ->options(Project::pluck('name', 'id'))
                        ->required(),

                    TinyEditor::make('description')
                        ->required(),

                    SpatieMediaLibraryFileUpload::make('attachments')
                        ->multiple()
                        ->collection('tasks'),

                     //...
                ])
                ->columns(2)
        ])
        ->model(Task::class)
        ->statePath('taskData');
}

public function saveTaskForm()
{
    $task = $this->task->update($this->taskForm->getState());

    $this->taskForm->model($task)->saveRelationships();

    $this->dispatchBrowserEvent('close-modal', ['id' => 'modalTask']);
}
```

**Comment by danharrin** [comment]
When you're editing the task, and you have a `->model()` passed to the taskForm, the form will save relationships before `getState()` is returned. So you dont need that line to save relationships. You also dont need the line before `fill()`, as the form should ALWAYS have access to the model, not just on the first livewire request when mount() is run.

**Reply by studiowizjo** [reply]
When I removed this 2 lines, form still works, but the main problem exists, so: `SpatieMediaLibraryFileUpload::make('attachments')` field is not populated with previously saved data and also new uploaded files are not saved.

**Reply by danharrin** [reply]
Please create a repo and I'll fix it and send it back

**Reply by studiowizjo** [reply]
I have just send you invitation to it: https://github.com/studiowizjo/filament-custom-form

I tried to simplify things as possible and make fresh installation with only essential data.

Steps to reproduce:
1. Run: `php artisan db:seed`
2. Go to: `admin/tasks/project-tasks`
3. Click link `Edit task #1`. Form will open with all data. You can change everyting and successfully save, besides attachments.

Thanks in advance for help =).

**Reply by danharrin** [reply]
Hello! Just tried this out, it worked first try from within the admin panel. I thought this was a problem with custom forms?

**Reply by studiowizjo** [reply]
Indeed, it is a problem with custom form. Have you tried it on page I described above: `admin/tasks/project-tasks`?

Here are the files we are talking about:
1. https://github.com/studiowizjo/filament-custom-form/blob/main/app/Filament/Resources/TaskResource/Pages/Traits/HasTasks.php
2. https://github.com/studiowizjo/filament-custom-form/blob/main/resources/views/filament/resources/tasks/project-tasks.blade.php

**Reply by danharrin** [reply]
Thank you for your patience.

I've committed a fix to the repo - the form needed a `model()` passed so that it has access to the task to fill and save images to.

I've also committed a second time, cleaning up code that didnt need to be there

Also, I don't think you need to extend ManageRecords? Only Page.

**Reply by studiowizjo** [reply]
Thank you very much for involvement. Everything works, but the strangest things is that when I want to apply changes to actual project, everything there was done exactly the same way and... now worked without any changes. I cannot understand what happened and could be the reason of that =/. It looks like some changes need to mature like wine.

Speaking about extending `ManageRecords` - the reason was I was  using table actions, but now, when custom forms works, there is no need.

Anyway, once again big thanks =).

**Reply by hrvoj3e** [reply]
Is there a snippet you could share that shows the solution for _"I've committed a fix to the repo - the form needed a model() passed so that it has access to the task to fill and save images to."_?

**Reply by studiowizjo** [reply]
```php
public Task $task;

public $taskData = [
    'id' => null,
    'project_id' => null,
    'user_id' => null,
    //...
];

public function openTaskModal(Task $task)
{
    $task->load('media');

    $this->task = $task;

    $this->taskForm
        ->model($this->task)
        ->fill($this->task->toArray());

    $this->dispatchBrowserEvent('open-modal', ['id' => 'modalTask']);
}

public function getTaskForm()
{
    return $this->makeForm()
        ->schema([
            Grid::make()
                ->schema([
                    //...
                ])
                ->columns(2)
        ])
        ->model($this->task ?? Task::class)
        ->statePath('taskData');
}

public function saveTaskForm()
{
    $this->task->update($this->taskForm->getState());

    $this->notify('success', 'Updated!');
    $this->dispatchBrowserEvent('close-modal', ['id' => 'modalTask']);
}
```
